packages/playground/blueprints/src/lib/compile.ts

```diff
@@ -171,33 +191,81 @@ export function compileBlueprint(
 			networking: blueprint.features?.networking ?? false,
 		},
 		run: async (playground: UniversalPHP) => {
+			const progressStatus = new Map<StepDefinition, Promise<void>>();
+			const progressPromises = new Map<
+				StepDefinition,
+				{
+					resolve: (result: any) => void;
+					reject: (error: any) => void;
+				}
+			>();
+			for (const { step } of compiled) {
+				const promise = new Promise<void>((resolve, reject) => {
+					progressPromises.set(step, { resolve, reject });
+				});
+				progressStatus.set(step, promise);
+				onStepProgress(promise, step);
+			}
+
 			try {
 				// Start resolving resources early
-				for (const { resources } of compiled) {
+				for (const { resources, step } of compiled) {
+					const stepPromise = progressPromises.get(step);
 					for (const resource of resources) {
 						resource.setPlayground(playground);
 						if (resource.isAsync) {
-							resource.resolve();
+							resource.resolve().catch(function (error) {
+								stepPromise?.reject(error);
+							});
 						}
 					}
 				}
 
+				/**
+				 * When a compiled step's `run` method is called, it
+				 * uses `resolveArguments` to await dependent resources.
+				 */
 				for (const { run, step } of compiled) {
-					const result = await run(playground);
-					onStepCompleted(result, step);
+					const stepPromise = progressPromises.get(step);
+					try {
+						const result = await run(playground);
+						stepPromise?.resolve(result);
+						onStepCompleted(result, step);
```

Comment:
If the `onStepCompleted` handler throws an error, the `stepPromise?.reject(error);` will run. It would be nice to just rethrow that error so the API consumer may notice and correct it.